docs/proposals/001-in-place-upgrades.md

```diff
@@ -0,0 +1,308 @@
+
+This operation should not be performed if `install.sh` script is overridden by the user in the manifests.
+
+This would prevent adding nodes with an outdated version and possibly breaking the cluster due to a version mismatch.
+
+## ControlPlane Provider Changes
+<!--
+This section MUST mention any changes to the controlplane provider.
+-->
+none
+
+## Configuration Changes
+<!--
+This section MUST mention any new configuration options or service arguments
+that are introduced.
+-->
+none
+
+## Documentation Changes
+<!--
+This section MUST mention any new documentation that is required for the new
+feature. Most features are expected to come with at least a How-To and an
+Explanation page.
+
+In this section, it is useful to think about any existing pages that need to be
+updated (e.g. command outputs).
+-->
+`How-To` page on performing in-place upgrades should be created.
+
+`Reference` page listing the annotations and possible values should be created/updated.
+
+## Testing
+<!--
+This section MUST explain how the new feature will be tested.
+-->
+The new feature can be tested manually by applying an annotation on the machine/node and checking for the `v1beta2.k8sd.io/in-place-upgrade-status` annotation's value to be `done`. A timeout should be set for waiting on the upgrade process.
+
+The tests can be integrated into the CI the same way with the CAPD infrastructure provider.
+
+The upgrade should be performed with the `localPath` option. Under Pebble the process would replace the `kubernetes` binary with the binary provided in the annotation value.
+
+This means a docker image containing 2 versions should be created. The different/new version of the `kubernetes` binary would also be built and put into a path.
+
+## Considerations for backwards compatibility
+<!--
+In this section, you MUST mention any breaking changes that are introduced by
+this feature. Some examples:
+
+- In case of deleting a database table, how do older k8sd instances handle it?
+- In case of a changed API endpoint, how do existing clients handle it?
+- etc
+-->
+
+## Implementation notes and guidelines
+<!--
+In this section, you SHOULD go into detail about how the proposal can be
+implemented. If needed, link to specific parts of the code (link against
+particular commits, not branches, such that any links remain valid going
+forward).
+
+This is useful as it allows the proposal owner to not be the person that
+implements it.
+-->
+
+The annotation method is chosen due to the "immutable infrastructure" assumption CAPI currently has. Which means updates are always done by creating new machines and fields are immutable. This might also pose some challenges on displaying accurate Kubernetes version information through CAPI.
```

Comment:
I think now the `version` column in `kubectl get machines` is not accurate, right? do we have any solutions for that? or is it acceptable?

Reply:
That is correct, sadly the `version` field is used in upstream controllers for reconciling machine sets. Setting that field has a lot of complexity, I believe it is acceptable to have inaccurate information there. IIRC the `kubectl describe machine` lists component versions for proxy, kubelet etc. which shows a correct value.

Reply:
got it, thanks a lot berkay!